crossbeam-epoch/src/storage/array_storage.rs

```diff
@@ -0,0 +1,247 @@
+/// appropriate `N`.
+///
+///
+/// # Ownership
+///
+/// `Array` is just a storage and doesn't own any elements inside the array.  In particular, it
+/// doesn't call `drop()` for its elements.
+///
+/// # Lifetime management
+///
+/// Because of non-standard memory layout, `Array` doesn't provide an ordinary constructor or
+/// destroyer.  Instead, the lifetime of an array is always managed by its owning pointer of type
+/// [`ArrayBox`].  See [`ArrayBox::new`] and [`ArrayBox::drop`] for more details.
+///
+/// [`ArrayBox`]: struct.ArrayBox.html
+/// [`ArrayBox::new`]: struct.ArrayBox.html#method.new
+/// [`ArrayBox::drop`]: struct.ArrayBox.html#impl-Drop
+#[derive(Debug)]
+pub struct Array<T> {
+    anchor: ManuallyDrop<T>,
+}
+
+impl<T> Array<T> {
+    /// Returns its size.
+    pub fn size(&self) -> usize {
+        let usize_align = mem::align_of::<usize>();
+        let usize_size = mem::size_of::<usize>();
+        let t_align = mem::align_of::<T>();
+        let t_size = mem::size_of::<T>();
+
+        unsafe {
+            // The memory layout varies depending on whether `T`'s alignment is <= `usize`'s
+            // alignment or not.
+            if t_align <= usize_align {
+                // Size is located a word before the anchor.
+                let ptr_num = (&self.anchor as *const _ as *const usize).sub(1);
+                ptr::read(ptr_num)
+            } else {
+                // Size is located `usize_elts * sizeof(T)`-bytes before the anchor.
+                let usize_elts = div_ceil(usize_size, t_size);
+                let ptr_num =
+                    (&self.anchor as *const ManuallyDrop<T>).sub(usize_elts) as *const usize;
+                ptr::read(ptr_num)
+            }
+        }
+    }
+
+    /// Returns the pointer to `index`-th element.
+    ///
+    /// # Safety
+    ///
+    /// `index` should be less than its size.  Otherwise, the behavior is undefined.
+    pub unsafe fn at(&self, index: usize) -> *const ManuallyDrop<T> {
+        debug_assert!(
+            index < self.size(),
+            "Array::at(): index {} should be < size {}",
+            index,
+            self.size()
+        );
+
+        let anchor = &self.anchor as *const ManuallyDrop<T>;
+        anchor.add(index)
+    }
+}
+
+/// The storage type for an [`Array`].
+///
+///
+/// See [`Array`] for more details.
+///
+/// # Examples
+///
+/// ```
+/// use crossbeam_epoch::{self as epoch, OwnedArray, Array, ArrayBox};
+///
+/// let a = ArrayBox::<i32>::new(10);
+/// let o = OwnedArray::from(a);
+/// ```
+///
+/// [`Array`]: struct.Array.html
+#[derive(Debug)]
+pub struct ArrayBox<T> {
+    ptr: *mut Array<T>,
+    _marker: PhantomData<T>,
+}
+
+fn div_ceil(a: usize, b: usize) -> usize {
+    (a + b - 1) / b
+}
+
+impl<T> ArrayBox<T> {
+    /// Creates a new array and returns the owning pointer to the new array.
+    pub fn new(num: usize) -> Self {
```

Comment:
We should really switch to using [`alloc`](https://doc.rust-lang.org/nightly/std/alloc/fn.alloc.html) and [`Layout`](https://doc.rust-lang.org/nightly/std/alloc/struct.Layout.html), which became stable in Rust 1.28.
Then the whole function simplifies to just a few lines of code.

Reply:
Thanks :)  Now it's much simpler!